packages/block-library/src/embed/edit.js

```diff
@@ -0,0 +1,382 @@
+			const { preview } = this.props;
+			const { url } = this.props.attributes;
+
+			if ( ! url ) {
+				return false;
+			}
+
+			const matchingBlock = findBlock( url );
+
+			// WordPress blocks can work on multiple sites, and so don't have patterns,
+			// so if we're in a WordPress block, assume the user has chosen it for a WordPress URL.
+			if ( WORDPRESS_EMBED_BLOCK !== this.props.name && DEFAULT_EMBED_BLOCK !== matchingBlock ) {
+				// At this point, we have discovered a more suitable block for this url, so transform it.
+				if ( this.props.name !== matchingBlock ) {
+					this.props.onReplace( createBlock( matchingBlock, { url } ) );
+					return true;
+				}
+			}
+
+			if ( preview ) {
+				const { html } = preview;
+
+				// We can't match the URL for WordPress embeds, we have to check the HTML instead.
+				if ( isFromWordPress( html ) ) {
+					// If this is not the WordPress embed block, transform it into one.
+					if ( WORDPRESS_EMBED_BLOCK !== this.props.name ) {
+						this.props.onReplace(
+							createBlock(
+								WORDPRESS_EMBED_BLOCK,
+								{
+									url,
+									// By now we have the preview, but when the new block first renders, it
+									// won't have had all the attributes set, and so won't get the correct
+									// type and it won't render correctly. So, we work out the attributes
+									// here so that the initial render works when we switch to the WordPress
+									// block. This only affects the WordPress block because it can't be
+									// rendered in the usual Sandbox (it has a sandbox of its own) and it
+									// relies on the preview to set the correct render type.
+									...this.getAttributesFromPreview(
+										this.props.preview, this.props.attributes.allowResponsive
+									),
+								}
+							)
+						);
+						return true;
+					}
+				}
+			}
+
+			return false;
+		}
+
+		/**
+		 * Gets the appropriate CSS class names to enforce an aspect ratio when the embed is resized
+		 * if the HTML has an iframe with width and height set.
+		 *
+		 * @param {string} html The preview HTML that possibly contains an iframe with width and height set.
+		 * @param {boolean} allowResponsive If the classes should be added, or removed.
+		 * @return {Object} Object with classnames set for use with `classnames`.
+		 */
+		getAspectRatioClassNames( html, allowResponsive = true ) {
+			const previewDocument = document.implementation.createHTMLDocument( '' );
+			previewDocument.body.innerHTML = html;
+			const iframe = previewDocument.body.querySelector( 'iframe' );
+
+			if ( iframe && iframe.height && iframe.width ) {
+				const aspectRatio = ( iframe.width / iframe.height ).toFixed( 2 );
+				// Given the actual aspect ratio, find the widest ratio to support it.
+				for ( let ratioIndex = 0; ratioIndex < ASPECT_RATIOS.length; ratioIndex++ ) {
+					const potentialRatio = ASPECT_RATIOS[ ratioIndex ];
+					if ( aspectRatio >= potentialRatio.ratio ) {
+						return {
```

Comment:
I think this could be:
```
return classnames( {
	[ potentialRatio.className ]: allowResponsive,
	'wp-has-aspect-ratio': allowResponsive,
} );
```
so that the function only returns a string. I think when another coder sees a function called getClassNames there would generally be an assumption that it returns a string useable as a class name.

Reply:
Yeah, we'd need to have it use `this.props.attributes.className` to calculate the final state of the attribute, because this is also used to remove the classes when the user toggles responsiveness for the block. I think there's some duplicate code in here too, so the PR that addresses this might have a More Red Than Green award :D

Reply:
Hmm, yeah, it could just be called `getClassNames` in that case.